docs/architecture.md

```diff
@@ -0,0 +1,465 @@
+    fn insert_consensus_data(sealed_block: &SealedFuelBlock) -> Result<()>;
+}
+
+trait PeerToPeer {
+    // broadcast a newly produced block to the network
+    async fn broadcast_new_block(block: SealedFuelBlock) -> Result<()>;
+}
+```
+#### Adapters: fuel_core::service::adapters::poa_consensus
+```rust
+impl fuel_sync::ports::Consensus for Service<PoA> {
+    // insert impl here
+}
+```
+
+### Block Producer
+
+#### Ports: fuel_block_producer::ports
+
+```rust
+trait TransactionPool {
+    fn select_block_txs(max_gas: u64, block_height: BlockHeight) -> Result<Vec<Arc<CheckedTransaction>>>;
+    fn drop_txs(txs: Vec<TxId>) -> Result<()>;
+}
+
+trait Executor {
+    async fn produce_block(block: &mut FuelBlock) -> Result<()>;
+}
+
+trait Relayer {
+    fn get_finalized_da_height() -> Result<DaHeight>;
+}
+```
+#### Adapters: fuel_core::service::adapters::block_producer
+```rust
+impl fuel_poa_consensus::ports::BlockProducer for Service<BlockProducer> {
+    // insert impl here
+}
+```
+
+### Block Importer
+
+#### Ports: fuel_block_importer::ports
+
+```rust
+trait Relayer {
+    // wait until the relayer is synced with ethereum
+    async fn await_synced() -> Result<()>;
+    // wait until a specific da_height is reached
+    async fn await_da_height(da_height: DaHeight) -> Result<()>;
+}
+
+trait Executor {
+    async fn validate_and_store_block(block: &FuelBlock) -> Result<()>;
+}
+
+trait TransactionPool {
+    // remove a set of txs from the pool after a block is committed
+    fn drop_txs(txs: Vec<TxId>) -> Result<()>;
+}
+
+trait Database {
+    // insert consensus information associated with a sealed block
+    fn insert_consensus_data(sealed_block: &SealedFuelBlock) -> Result<()>;
+}
+
+```
+#### Adapters: fuel_core::service::adapters::block_importer
+```rust
+impl fuel_poa_consensus::ports::BlockImporter for Service<BlockImporter> {
+    // insert impl here
+}
+
+impl fuel_sync::ports::BlockImporter for Service<BlockImporter> {
+    // insert impl here
+}
+```
+
+### Synchronizer
+
+#### Ports: fuel_sync::ports
+
+```rust
+trait PeerToPeer {
+    type SealedHeaderResponse: NetworkData<SealedFuelBlockHeader>;
+    type BlockResponse: NetworkData<Vec<SealedFuelBlock>>;
+    type GossipedBlock: NetworkData<SealedFuelBlock>;
+    
+    async fn fetch_best_network_block_header() -> Result<Self::SealedHeaderResponse>;
+    async fn fetch_blocks(query: Range<BlockHeight>) -> Result<Self::BlockResponse>;
+    // punish the sender for providing an invalid block header
+    fn report_invalid_block_header(invalid_header: &Self::SealedHeaderResponse) -> Result<()>;
+    // punish the sender for providing a set of blocks that aren't valid
+    fn report_invalid_blocks(invalid_blocks: &Self::BlockResponse) -> Result<()>;
+    // await a newly produced block from the network (similar to stream.next())
+    async fn next_gossiped_block() -> Result<Self::GossipedBlock>;
```

Comment:
Do we really want to gossip about the whole `FuelBlock`? And handle it in another channel?

Other peers can gossip about the new height via `fetch_best_network_block_header` and then we can use default logic to process it.

PoA peer -> produce block -> commit block locally -> gossip about new height -> `fetch_best_network_block_header` on non-PoA node -> `fetch_blocks` with new height -> await new block and start process it.

Reply:
Ah true - the peer might not care about the gossiped message and so it would be wasteful to include the full block.

The only place where we might want to consider broadcasting the full block is during consensus (i.e. POS), to cut down on latency.

Reply:
Yep, but for PoS, we will have a separate channel/stream to receive only blocks/votes related for consensus. FS should not be aware of this data=)

Reply:
Ah so you're saying we don't need gossip at all and can just poll for headers?

Reply:
Yeah the full block for consensus would be a separate kind of message than what the synchronizer would subscribe to.

Reply:
> Ah so you're saying we don't need gossip at all and can just poll for headers?

Yep, non-PoA peers will receive a block from the synchronizer by default synchronization flow=) In this case we don't need to write any additional logic

Reply:
So in this case we just remove the PoA sealed block broadcast altogether? I suppose the liveness shouldn't be too terrible assuming our block times are higher than the default synchronizer polling interval.

Reply:
Yep, block header polling/gossiping should be enough for syncing=)

Reply:
Given this simplification, how about we continue letting block production save to db and the BI commit (which validates blocks) also save to db? The only overlap between POA and BI at that point would be persisting the consensus metadata and the system would be more efficient.

When we implement PoS later, it would have to define it's own block producer port which could return a state_diff instead of immediately persisting. This way we keep the PoA mode efficient, without trying to plan too far ahead and by attempting to align with interfaces/usage patterns only needed for PoS.

Reply:
I meant we don't need to broadcast the whole block, only the header=D It decreases latency for the network to sync=) Also, reduce the number of polling=D

But yea, we can to not broadcast it to the network=) But still, we need to notify the transaction pool, and Block Importer is a good place for it=)

Reply:
Sorry for the misunderstanding, broadcasting the finalized block headers makes sense.

Reply:
ok updated to gossip headers instead of full blocks, keeping the gossipsync flow intact.

also updated the poa to use the imported for committing the finalized state of the block to avoid multiple committers.